### ISSUE TITLE
[Feature Request] Human-readable exception messages for incorrectly chained decorators

### ISSUE THREAD

**Issue by qiujiangkun** [beartype/beartype#80]
```python
from beartype import beartype
class Bar:
    @beartype
    @classmethod
    def foo(cls, s: str) -> int:
        return int(s)
```
gives
<classmethod object at 0x7f68d6e9fbb0> uncallable

**Comment by leycec** [comment]
**Ah-ha!** You're in luck in 2022, because `beartype` already supports what you want. You just need to reverse the order in which you're chaining decorators: e.g.,

``` python
from beartype import beartype
class Bar:
    @classmethod
    @beartype
    def foo(cls, s: str) -> int:
        return int(s)
```

Of course, the same also applies to other builtin decorators like `@staticmethod`, `@property`, and `@{property_name}.setter`. See [the `class MyOtherClass:` example in our voluminous cheatsheet for additional commentary](https://github.com/beartype/beartype#cheatsheet).

You're *right*, though. `beartype` should explicitly tell you this in its exception. Let's edit your title and leave this issue open to remind me to eventually do this.

Happy New Year, Qiu. May the Year of the Tiger bless us all with type-safe code. :tiger:

**Comment by qiujiangkun** [comment]
How about extracting the underlying function and re-annotate with classmethod and staticmethod? I'm currently using this approch. But it doesn't work for `@property`
```python

if isinstance(value, classmethod):
    attr[name] = classmethod(beartype(value.__func__))
elif isinstance(value, staticmethod):
    attr[name] = staticmethod(beartype(value.__func__))
```

**Comment by TeamSpen210** [comment]
For `property` there are potentially 3 functions you need to wrap - `.fget`, `.fset` and `.fdel`. If not set they're `None`, so you'll need to not wrap in that case. Once you've wrapped them pass to the `property` constructor, plus `doc=old.__doc__` to keep whatever was chosen.

**Comment by leycec** [comment]
> How about extracting the underlying function and re-annotate with `classmethod` and `staticmethod`?

:face_exhaling: 

> For `property` there are potentially 3 functions you need to wrap - `.fget`, `.fset` and `.fdel`.

:face_with_spiral_eyes: 

You're both absolutely right, of course. We can (and possibly should) be clever here and just do what the user meant by forcibly beating Python into submission. I'm all for beatings.

But... I'm kinda concerned about unanticipated dragons. I can feel their hot breath weapons on my exposed neck – especially that eldritch level 99 black dragon called `@property`. That's another level of darkness. I don't want to break Microsoft, because then Microsoft breaks me.

## @leycec Presents a Hideous Dragon

I actually *just* hit a subtle dragon under Python ≥ 3.10 while resolving this tonight:

``` python
$ ipython3.10
Python 3.10.1 (main, Dec 17 2021, 03:20:21) [GCC 10.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.
>>> callable(staticmethod(lambda cls: cls.__name__))
True
```

What the codswallop is this!?!? **Yup.** It is all too true.

`@staticmethod` descriptors now define the `__call__()` dunder method and are thus directly callable under Python ≥ 3.10. Of course, they're *not* pure-Python functions; they're C-based. Since they don't provide code objects or annotations, we still can't directly decorate them; they're just harder to detect. Thankfully, they *do* still define the `__func__` dunder variable whose value is the original undecorated pure-Python function.

*Annnyway* – the point is that even reliably detecting these things is becoming non-trivial across Python versions. I'm also concerned about:

* **PyPy** and how PyPy implements these descriptors, 'cause PyPy always gotta be different and break everything.
* Static code analysis, especially static type checkers like **mypy**. Why? Because...

Even if we fix up every disordered decorator chain at runtime, static analysis tools will *still* perceive `@beartype` as erroneously decorating C-based descriptor objects from their perspective. Will a future version of mypy complain?

**You bet your bottom dollar.** I don't know what that idiom means, but I *do* know mypy is anal-retentive and obsessive-compulsive to a fault. That's not a bad thing; that's why we all use mypy in the first place. But it does mean that mypy can and will raise errors about that unexpected behaviour at some point (if it doesn't already), which then means that we'd be inviting [PEP 561 violations](https://www.python.org/dev/peps/pep-0561).

## But What If Everything's Just Peachy, Huh?

Let's assume there are no dragons.

Even then, `@beartype` should still at least emit non-fatal warnings on detecting and fixing these disordered decoration chains. Why? Because fixing them in `@beartype` will incur a minor performance penalty at decoration time... usually paid at application startup. It's preferable (from pretty much every perspective) for users to resolve this on their end – even if we can (and possibly should) lend them a helping hand.

For now, commit fbeae42 assumes the lazy posture by raising human-readable exceptions instructing users on what to do:

``` python
>>> from beartype import beartype
>>> class MuhClass(object):
...     @beartype
...     @staticmethod
...     def muh_class_method(cls) -> str:
...         return cls.__name__
beartype.roar.BeartypeDecorWrappeeException: Uncallable 
descriptor created by builtin decorator @staticmethod not
decoratable by @beartype. Consider listing @beartype below
rather than above @staticmethod in the decorator chain for
this method: e.g.,
	@staticmethod
	@beartype      # <-- this is the Way of the Bear
	def ...
```

I acknowledge this is cowardly. Since you're both right, let's reopen this later. Helpfully pester me about this in a month or two and I'll have a go at fixing up at least disordered `@classmethod` and `@staticmethod` decoration chains.
`@property` still frightens me.

For now, I hope no one objects if I momentarily close this. The number of open issues is exploding uncontrollably. But... how can this be!?!? We're only on Day 2 of 2022 here! `</gulp>`

**Comment by leycec** [comment]
**Resurrect this issue,** GitHub! @qiujiangkun and @TeamSpen210: you were right about everything.

@beartype *should* just permissively decorate everything it possibly can, because that is always the right thing to do. The wrong thing to do is what @beartype currently does – which is to force everyone to chain their decorators in an arbitrary way to satisfy my personal purist ideology. Bad news, @beartype.

More importantly, this issue has now become a roadblock to resolving an even more critical issue: **[class decoration](https://github.com/beartype/beartype/issues/152).** The only sane way for @beartype to decorate arbitrary user-defined classes like this...

```python
from beartype import beartype

@beartype
def MuhClass(object):
    @property
    def muh_property(self) -> str:
        return 'This is my property. There are many like it. This is mine.'

    @classmethod
    def muh_classmethod() -> str:
        return 'The journey of a thousand bugs begins today.'

    @staticmethod
    def muh_staticmethod() -> str:
        return 'Let he who is without bugs cast the first issue.'
```

...is to generalize @beartype to support decorating properties, class methods, and static methods in any arbitrary decorator order. Everyone was right. I'm bearly humble enough to admit that now. :face_exhaling:

**Comment by leycec** [comment]
**Resolved by bd69ec0.** Behold! Glory approaches:

```python
>>> from beartype import beartype
>>> class Bar:
...     @beartype
...     @classmethod
...     def foo(cls, s: str) -> int:
...         return int(s)
>>> print(f'{Bar().foo("1337")}')
1337
```

:partying_face:
